File: adidas-scraper/browser/python/adidas.py
"""Adidas scraper using the official Scrapeless Python SDK + Playwright over CDP.

Surfaces:
- scrape_product(urls)         -> list of Product dicts (PDP fields lifted from JSON-LD + DOM)
- scrape_search(url, max_pages) -> list of SearchResult dicts (PLP product cards)

Under the hood:
- `client.browser.create()` mints a cloud session on Scrapeless's Scraping Browser,
  returning a CDP WebSocket endpoint (`browser_ws_endpoint`).
- Playwright connects over CDP, drives the page, returns rendered HTML.
- Parsel parses the HTML; we prefer JSON-LD blobs over CSS classes when possible.

Adidas.com is protected by Akamai Bot Manager so retries + Scrapeless's residential
fingerprinting matter; we ship `retries=2` by default and wait on a stable marker.
"""

from __future__ import annotations

import json
import os
import re
from typing import Any, Dict, List, Optional
from urllib.parse import urljoin, urlparse, parse_qsl, urlencode, urlunparse

from loguru import logger
from parsel import Selector
from playwright.async_api import async_playwright
from scrapeless import Scrapeless
from scrapeless.types import ICreateBrowser

DEFAULT_PROXY_COUNTRY = "US"
DEFAULT_SESSION_TTL = 180

# ---------------------------------------------------------------------------
# Scrapeless plumbing
# ---------------------------------------------------------------------------

def _client() -> Scrapeless:
    if not os.environ.get("SCRAPELESS_API_KEY") and not os.environ.get("SCRAPELESS_KEY"):
        raise RuntimeError(
            "SCRAPELESS_API_KEY env var not set. Sign up at https://app.scrapeless.com"
        )
    if not os.environ.get("SCRAPELESS_API_KEY") and os.environ.get("SCRAPELESS_KEY"):
        os.environ["SCRAPELESS_API_KEY"] = os.environ["SCRAPELESS_KEY"]
    return Scrapeless()

async def _fetch_rendered_html(
    url: str,
    ready_selector: Optional[str] = None,
    *,
    proxy_country: str = DEFAULT_PROXY_COUNTRY,
    retries: int = 2,
) -> str:
    """Discover -> extract: mint a session, goto, wait for stable marker, return HTML."""
    last_error: Optional[Exception] = None
    for attempt in range(retries + 1):
        client = _client()
        session = client.browser.create(
            ICreateBrowser(proxy_country=proxy_country, session_ttl=DEFAULT_SESSION_TTL)
        )
        async with async_playwright() as pw:
            browser = await pw.chromium.connect_over_cdp(session.browser_ws_endpoint)
            try:
                page = await browser.new_page()
                await page.goto(url, wait_until="domcontentloaded", timeout=60000)
                if ready_selector:
                    try:
                        await page.wait_for_selector(ready_selector, timeout=20000)
                    except Exception as e:
                        logger.warning("wait_for_selector {!r} failed (continuing): {}", ready_selector, e)
                html = await page.content()
                if html and "<html" in html.lower():
                    return html
                last_error = RuntimeError("empty HTML")
            except Exception as e:  # noqa: BLE001
                last_error = e
                logger.warning("attempt {} failed: {}", attempt + 1, e)
            finally:
                try:
                    await browser.close()
                except Exception:
                    pass
    raise RuntimeError(f"giving up after {retries + 1} attempts: {last_error}")

# ---------------------------------------------------------------------------
# JSON-LD helpers
# ---------------------------------------------------------------------------

def _iter_jsonld_nodes(raw_blocks: List[str]):
    for raw in raw_blocks:
        if not raw or not raw.strip():
            continue
        try:
            data = json.loads(raw)
        except json.JSONDecodeError:
            continue
        nodes = data if isinstance(data, list) else [data]
        for node in nodes:
            if not isinstance(node, dict):
                continue
            graph = node.get("@graph")
            if isinstance(graph, list):
                for sub in graph:
                    if isinstance(sub, dict):
                        yield sub
            else:
                yield node

def _type_matches(node: Dict[str, Any], wanted: str) -> bool:
    t = node.get("@type")
    if isinstance(t, str):
        return t == wanted
    if isinstance(t, list):
        return wanted in t
    return False

def _first_offer(node: Dict[str, Any]) -> Dict[str, Any]:
    offers = node.get("offers")
    if isinstance(offers, list) and offers:
        return offers[0] if isinstance(offers[0], dict) else {}
    if isinstance(offers, dict):
        nested = offers.get("offers")
        if isinstance(nested, list) and nested:
            return nested[0] if isinstance(nested[0], dict) else {}
        return offers
    return {}

def _aggregate_rating(node: Dict[str, Any]) -> Dict[str, Any]:
    ar = node.get("aggregateRating")
    return ar if isinstance(ar, dict) else {}

# ---------------------------------------------------------------------------
# Parsers
# ---------------------------------------------------------------------------

_PRODUCT_ID_RE = re.compile(r"/([A-Z]{2}\d{4}|[A-Z]{3}\d{2}|[A-Z]{2}\d{3}[A-Z]?|[A-Z]{1,3}\d{3,4})\.html", re.I)

def _extract_product_id(url: str) -> str:
    m = _PRODUCT_ID_RE.search(url)
    if m:
        return m.group(1).upper()
    # fallback: last path segment minus .html
    path = urlparse(url).path.rstrip("/")
    last = path.split("/")[-1]
    return last.replace(".html", "")

def _clean(value: Optional[str]) -> Optional[str]:
    if value is None:
        return None
    v = re.sub(r"\s+", " ", value).strip()
    return v or None

def parse_product(html: str, url: str) -> Dict[str, Any]:
    """Parse an Adidas PDP. Prefers JSON-LD Product; falls back to DOM."""
    sel = Selector(text=html)
    raw_blocks = sel.css('script[type="application/ld+json"]::text').getall()

    ld: Dict[str, Any] = {}
    for node in _iter_jsonld_nodes(raw_blocks):
        if _type_matches(node, "Product"):
            ld = node
            break

    offer = _first_offer(ld)
    rating = _aggregate_rating(ld)

    images_raw = ld.get("image") if ld else None
    if isinstance(images_raw, str):
        images = [images_raw]
    elif isinstance(images_raw, list):
        images = [str(x) for x in images_raw if x]
    else:
        images = sel.css('meta[property="og:image"]::attr(content)').getall()

    name = ld.get("name") if ld else None
    if not name:
        name = _clean(sel.css("h1::text").get())

    description = ld.get("description") if ld else None
    if not description:
        description = _clean(sel.css('meta[name="description"]::attr(content)').get())

    sku = ld.get("sku") or ld.get("productID") or _extract_product_id(url)

    brand_node = ld.get("brand") if ld else None
    if isinstance(brand_node, dict):
        brand = brand_node.get("name")
    elif isinstance(brand_node, str):
        brand = brand_node
    else:
        brand = "adidas"

    price_value: Optional[float] = None
    raw_price = offer.get("price") if offer else None
    if raw_price is not None:
        try:
            price_value = float(str(raw_price).replace(",", ""))
        except (TypeError, ValueError):
            price_value = None

    currency = offer.get("priceCurrency") if offer else None

    availability = offer.get("availability") if offer else None
    if isinstance(availability, str) and "/" in availability:
        availability = availability.rsplit("/", 1)[-1]

    return {
        "id": str(sku),
        "url": url,
        "name": _clean(name) or "",
        "brand": _clean(brand) or "adidas",
        "description": _clean(description),
        "price": price_value,
        "priceCurrency": currency,
        "availability": availability,
        "images": images,
        "rating": float(rating["ratingValue"]) if rating.get("ratingValue") else None,
        "reviewCount": int(rating["reviewCount"]) if rating.get("reviewCount") else None,
        "category": _clean(ld.get("category")) if ld else None,
        "color": _clean(ld.get("color")) if ld else None,
    }

def parse_search(html: str) -> Dict[str, Any]:
    """Parse an Adidas PLP / search page.

    Adidas's PLP injects card data into ItemList JSON-LD AND into the rendered DOM
    via `[data-testid="plp-product-card"]`. We prefer JSON-LD ItemList; fall back to DOM cards.
    """
    sel = Selector(text=html)
    raw_blocks = sel.css('script[type="application/ld+json"]::text').getall()

    items: List[Dict[str, Any]] = []
    for node in _iter_jsonld_nodes(raw_blocks):
        if not _type_matches(node, "ItemList"):
            continue
        for el in node.get("itemListElement") or []:
            if not isinstance(el, dict):
                continue
            item = el.get("item") if isinstance(el.get("item"), dict) else el
            offer = _first_offer(item)
            url = item.get("url") or el.get("url") or ""
            sku = item.get("sku") or item.get("productID") or _extract_product_id(url)
            price_value = None
            raw_price = offer.get("price") if offer else None
            if raw_price is not None:
                try:
                    price_value = float(str(raw_price).replace(",", ""))
                except (TypeError, ValueError):
                    price_value = None
            items.append({
                "id": str(sku) if sku else "",
                "url": url,
                "name": item.get("name") or "",
                "image": (item.get("image") if isinstance(item.get("image"), str)
                          else (item.get("image") or [None])[0] if isinstance(item.get("image"), list) else None),
                "price": price_value,
                "priceCurrency": offer.get("priceCurrency") if offer else None,
            })
        if items:
            break

    if not items:
        # DOM fallback - works for adidas.com PLPs rendered by glass UI
        for card in sel.css('[data-testid="plp-product-card"], article[data-testid="product-card"]'):
            link = card.css('a::attr(href)').get("") or ""
            absolute = urljoin("https://www.adidas.com", link) if link else ""
            sku = _extract_product_id(absolute) if absolute else (card.attrib.get("data-grid-id") or "")
            name = _clean(card.css('[data-testid="product-card-title"]::text').get()
                          or card.css('p::text').get())
            price_text = _clean(
                card.css('[data-testid="primary-price"]::text').get()
                or card.css('div[class*="price"] *::text').get()
            )
            price_value = None
            if price_text:
                m = re.search(r"[\d,.]+", price_text)
                if m:
                    try:
                        price_value = float(m.group(0).replace(",", ""))
                    except ValueError:
                        price_value = None
            img = (card.css('img::attr(src)').get()
                   or card.css('img::attr(data-src)').get())
            items.append({
                "id": sku or "",
                "url": absolute,
                "name": name or "",
                "image": img,
                "price": price_value,
                "priceCurrency": "USD" if price_text and "$" in price_text else None,
            })

    return {"results": items}

# ---------------------------------------------------------------------------
# URL helpers
# ---------------------------------------------------------------------------

def _add_query(url: str, **params: Any) -> str:
    parsed = urlparse(url)
    q = dict(parse_qsl(parsed.query))
    q.update({k: str(v) for k, v in params.items()})
    return urlunparse(parsed._replace(query=urlencode(q)))

# ---------------------------------------------------------------------------
# Scrape functions
# ---------------------------------------------------------------------------

async def scrape_product(urls: List[str]) -> List[Dict[str, Any]]:
    """Scrape one or more Adidas PDP URLs."""
    results: List[Dict[str, Any]] = []
    for url in urls:
        logger.info("scraping product {}", url)
        html = await _fetch_rendered_html(url, ready_selector='script[type="application/ld+json"]')
        results.append(parse_product(html, url))
    return results

async def scrape_search(url: str, max_pages: int = 1) -> List[Dict[str, Any]]:
    """Scrape an Adidas PLP / search URL and follow pagination (`?start=` offsets of 48)."""
    logger.info("scraping search {}", url)
    first_html = await _fetch_rendered_html(url, ready_selector='script[type="application/ld+json"]')
    parsed = parse_search(first_html)
    out: List[Dict[str, Any]] = list(parsed["results"])

    # Adidas uses ?start=48,96,... (page size 48) on most US PLPs
    page_size = 48
    for page in range(2, max_pages + 1):
        offset = (page - 1) * page_size
        page_url = _add_query(url, start=offset)
        page_html = await _fetch_rendered_html(page_url, ready_selector='script[type="application/ld+json"]')
        out.extend(parse_search(page_html)["results"])
    logger.success("search {} returned {} items", url, len(out))
    return out

def to_dict(obj: Any) -> Any:
    if isinstance(obj, list):
        return [to_dict(x) for x in obj]
    return obj

File: adidas-scraper/browser/nodejs/adidas.mjs
// Adidas scraper using the official @scrapeless-ai/sdk + puppeteer-core over CDP.
//
// Surfaces:
//   - scrapeProduct(urls)        -> array of Product dicts (PDP fields lifted from JSON-LD + DOM)
//   - scrapeSearch(url, maxPages) -> array of SearchResult dicts (PLP cards)
//
// Adidas.com is protected by Akamai; retries + Scrapeless's residential
// fingerprinting matter. We ship retries=2 by default.

import { Scrapeless } from "@scrapeless-ai/sdk";
import puppeteer from "puppeteer-core";
import * as cheerio from "cheerio";

const DEFAULT_PROXY_COUNTRY = "US";
const DEFAULT_SESSION_TTL = 180;

function requireKey() {
  const k = process.env.SCRAPELESS_API_KEY ?? process.env.SCRAPELESS_KEY;
  if (!k) throw new Error("SCRAPELESS_API_KEY env var not set. Sign up at https://app.scrapeless.com");
  return k;
}

function client() {
  return new Scrapeless({ apiKey: requireKey() });
}

async function fetchRenderedHtml(url, readySelector, { proxyCountry = DEFAULT_PROXY_COUNTRY, retries = 2 } = {}) {
  let lastError;
  for (let attempt = 0; attempt <= retries; attempt++) {
    const session = await client().browser.create({ proxyCountry, sessionTTL: DEFAULT_SESSION_TTL });
    let browser;
    try {
      browser = await puppeteer.connect({ browserWSEndpoint: session.browserWSEndpoint });
      const page = await browser.newPage();
      await page.goto(url, { waitUntil: "domcontentloaded", timeout: 60000 });
      if (readySelector) {
        try { await page.waitForSelector(readySelector, { timeout: 20000 }); } catch (_) {}
      }
      const html = await page.content();
      if (html && html.toLowerCase().includes("<html")) return html;
      lastError = new Error("empty HTML");
    } catch (e) {
      lastError = e;
    } finally {
      try { await browser?.close(); } catch (_) {}
    }
  }
  throw new Error(`giving up after ${retries + 1} attempts: ${lastError?.message}`);
}

// ---------- JSON-LD helpers ----------

function* iterJsonldNodes($) {
  const blocks = $('script[type="application/ld+json"]')
    .map((_, el) => $(el).contents().text())
    .get();
  for (const raw of blocks) {
    if (!raw || !raw.trim()) continue;
    let data;
    try { data = JSON.parse(raw); } catch { continue; }
    const nodes = Array.isArray(data) ? data : [data];
    for (const node of nodes) {
      if (!node || typeof node !== "object") continue;
      if (Array.isArray(node["@graph"])) {
        for (const sub of node["@graph"]) {
          if (sub && typeof sub === "object") yield sub;
        }
      } else {
        yield node;
      }
    }
  }
}

function typeMatches(node, wanted) {
  const t = node["@type"];
  if (typeof t === "string") return t === wanted;
  if (Array.isArray(t)) return t.includes(wanted);
  return false;
}

function firstOffer(node) {
  const offers = node?.offers;
  if (Array.isArray(offers) && offers.length) {
    return typeof offers[0] === "object" ? offers[0] : {};
  }
  if (offers && typeof offers === "object") {
    if (Array.isArray(offers.offers) && offers.offers.length) {
      return typeof offers.offers[0] === "object" ? offers.offers[0] : {};
    }
    return offers;
  }
  return {};
}

function aggregateRating(node) {
  const ar = node?.aggregateRating;
  return (ar && typeof ar === "object") ? ar : {};
}

// ---------- parsers ----------

const PRODUCT_ID_RE = /\/([A-Z]{2}\d{4}|[A-Z]{3}\d{2}|[A-Z]{2}\d{3}[A-Z]?|[A-Z]{1,3}\d{3,4})\.html/i;

function extractProductId(url) {
  if (!url) return "";
  const m = url.match(PRODUCT_ID_RE);
  if (m) return m[1].toUpperCase();
  try {
    const u = new URL(url);
    const last = u.pathname.replace(/\/$/, "").split("/").pop() || "";
    return last.replace(/\.html$/i, "");
  } catch { return ""; }
}

function clean(value) {
  if (value == null) return null;
  const v = String(value).replace(/\s+/g, " ").trim();
  return v || null;
}

export function parseProduct(html, url) {
  const $ = cheerio.load(html);

  let ld = {};
  for (const node of iterJsonldNodes($)) {
    if (typeMatches(node, "Product")) { ld = node; break; }
  }

  const offer = firstOffer(ld);
  const rating = aggregateRating(ld);

  let images = [];
  if (typeof ld.image === "string") images = [ld.image];
  else if (Array.isArray(ld.image)) images = ld.image.filter(Boolean).map(String);
  else {
    images = $('meta[property="og:image"]').map((_, el) => $(el).attr("content")).get().filter(Boolean);
  }

  const name = ld.name ? clean(ld.name) : clean($("h1").first().text());
  const description = ld.description
    ? clean(ld.description)
    : clean($('meta[name="description"]').attr("content"));

  const sku = ld.sku || ld.productID || extractProductId(url);

  let brand = "adidas";
  if (ld.brand) {
    brand = typeof ld.brand === "object" ? (ld.brand.name || "adidas") : String(ld.brand);
  }

  let priceValue = null;
  if (offer.price != null) {
    const n = parseFloat(String(offer.price).replace(/,/g, ""));
    priceValue = Number.isFinite(n) ? n : null;
  }

  let availability = offer.availability ?? null;
  if (typeof availability === "string" && availability.includes("/")) {
    availability = availability.split("/").pop();
  }

  return {
    id: String(sku || ""),
    url,
    name: name || "",
    brand: clean(brand) || "adidas",
    description,
    price: priceValue,
    priceCurrency: offer.priceCurrency ?? null,
    availability,
    images,
    rating: rating.ratingValue != null ? parseFloat(rating.ratingValue) : null,
    reviewCount: rating.reviewCount != null ? parseInt(rating.reviewCount, 10) : null,
    category: ld.category ? clean(ld.category) : null,
    color: ld.color ? clean(ld.color) : null,
  };
}

export function parseSearch(html) {
  const $ = cheerio.load(html);
  const items = [];

  for (const node of iterJsonldNodes($)) {
    if (!typeMatches(node, "ItemList")) continue;
    for (const el of node.itemListElement || []) {
      if (!el || typeof el !== "object") continue;
      const item = (el.item && typeof el.item === "object") ? el.item : el;
      const offer = firstOffer(item);
      const url = item.url || el.url || "";
      const sku = item.sku || item.productID || extractProductId(url);
      let priceValue = null;
      if (offer.price != null) {
        const n = parseFloat(String(offer.price).replace(/,/g, ""));
        priceValue = Number.isFinite(n) ? n : null;
      }
      let img = null;
      if (typeof item.image === "string") img = item.image;
      else if (Array.isArray(item.image) && item.image.length) img = String(item.image[0]);
      items.push({
        id: sku ? String(sku) : "",
        url,
        name: item.name || "",
        image: img,
        price: priceValue,
        priceCurrency: offer.priceCurrency ?? null,
      });
    }
    if (items.length) break;
  }

  if (!items.length) {
    $('[data-testid="plp-product-card"], article[data-testid="product-card"]').each((_, el) => {
      const $card = $(el);
      const link = $card.find("a").first().attr("href") || "";
      let absolute = "";
      try { absolute = link ? new URL(link, "https://www.adidas.com").toString() : ""; } catch {}
      const sku = absolute ? extractProductId(absolute) : ($card.attr("data-grid-id") || "");
      const name = clean(
        $card.find('[data-testid="product-card-title"]').first().text() ||
        $card.find("p").first().text()
      );
      const priceText = clean(
        $card.find('[data-testid="primary-price"]').first().text() ||
        $card.find('div[class*="price"]').first().text()
      );
      let priceValue = null;
      if (priceText) {
        const m = priceText.match(/[\d,.]+/);
        if (m) {
          const n = parseFloat(m[0].replace(/,/g, ""));
          priceValue = Number.isFinite(n) ? n : null;
        }
      }
      const img = $card.find("img").first().attr("src") || $card.find("img").first().attr("data-src") || null;
      items.push({
        id: sku || "",
        url: absolute,
        name: name || "",
        image: img,
        price: priceValue,
        priceCurrency: priceText && priceText.includes("$") ? "USD" : null,
      });
    });
  }

  return { results: items };
}

// ---------- URL helpers ----------

function addQuery(url, params) {
  const u = new URL(url);
  for (const [k, v] of Object.entries(params)) u.searchParams.set(k, String(v));
  return u.toString();
}

// ---------- scrape functions ----------

export async function scrapeProduct(urls) {
  const list = Array.isArray(urls) ? urls : [urls];
  const out = [];
  for (const url of list) {
    const html = await fetchRenderedHtml(url, 'script[type="application/ld+json"]');
    out.push(parseProduct(html, url));
  }
  return out;
}

export async function scrapeSearch(url, maxPages = 1) {
  const firstHtml = await fetchRenderedHtml(url, 'script[type="application/ld+json"]');
  const out = [...parseSearch(firstHtml).results];

  const pageSize = 48;
  for (let page = 2; page <= maxPages; page++) {
    const offset = (page - 1) * pageSize;
    const pageUrl = addQuery(url, { start: offset });
    const html = await fetchRenderedHtml(pageUrl, 'script[type="application/ld+json"]');
    out.push(...parseSearch(html).results);
  }
  return out;
}
